mirrord/operator/src/client.rs

```diff
@@ -1,184 +1,184 @@
+use crate::crd::TargetCrd;
 
 static CONNECTION_CHANNEL_SIZE: usize = 1000;
 
-pub struct OperatorApi<T: ToSocketAddrs> {
-    addr: T,
-    target: TargetConfig,
-}
-
-impl<T> OperatorApi<T>
-where
-    T: ToSocketAddrs,
-{
-    pub fn new(addr: T, target: TargetConfig) -> Self {
-        OperatorApi { addr, target }
-    }
+#[derive(Debug, Error)]
+pub enum OperatorApiError {
+    #[error("unable to create target for TargetConfig")]
+    InvalidTarget,
+    #[error(transparent)]
+    HttpError(#[from] http::Error),
+    #[error(transparent)]
+    KubeApiError(#[from] KubeApiError),
 }
 
-#[async_trait]
-impl<T> AgentManagment for OperatorApi<T>
-where
-    T: ToSocketAddrs + Send + Sync,
-{
-    type AgentRef = actix_codec::Framed<TcpStream, OperatorCodec>;
-    type Err = io::Error;
+type Result<T, E = OperatorApiError> = std::result::Result<T, E>;
 
-    async fn create_connection(
-        &self,
-        codec: Self::AgentRef,
-    ) -> Result<(mpsc::Sender<ClientMessage>, mpsc::Receiver<DaemonMessage>), Self::Err> {
-        wrap_connection(codec)
-    }
+pub struct OperatorApi {
+    client: Client,
+    target: TargetConfig,
+}
 
-    async fn create_agent<P>(&self, _: &P) -> Result<Self::AgentRef, Self::Err>
+impl OperatorApi {
+    pub async fn discover<P>(config: &LayerConfig, progress: &P) -> Result<(Self, TargetCrd)>
     where
         P: Progress + Send + Sync,
     {
-        let stream = TcpStream::connect(&self.addr).await?;
-
-        Ok(connection(stream, self.target.clone()).await)
-    }
-}
-
-pub struct OperatorApiDiscover {
-    client: Client,
-    operator: OperatorConfig,
-    target: TargetConfig,
-}
+        let operator_api = OperatorApi::new(&config).await?;
 
-impl OperatorApiDiscover {
-    pub async fn create(config: &LayerConfig) -> Result<Self, KubeApiError> {
-        let client = create_kube_api(Some(config.clone())).await?;
+        let operator_ref = operator_api.create_agent(progress).await?;
 
-        Ok(OperatorApiDiscover {
-            client,
-            operator: config.operator.clone(),
-            target: config.target.clone(),
-        })
+        Ok((operator_api, operator_ref))
     }
 
-    pub async fn discover_operator<P: Progress + Send + Sync>(
-        config: &LayerConfig,
-        progress: &P,
-    ) -> Option<(Self, <Self as AgentManagment>::AgentRef)> {
-        let api = OperatorApiDiscover::create(config).await.ok()?;
+    pub async fn new(config: &LayerConfig) -> Result<Self> {
+        let target = config.target.clone();
 
-        let connection = api.create_agent(progress).await.ok()?;
+        let client = create_kube_api(Some(config.clone())).await?;
 
-        Some((api, connection))
+        Ok(OperatorApi { client, target })
     }
 }
 
 #[async_trait]
-impl AgentManagment for OperatorApiDiscover {
-    type AgentRef = String;
-    type Err = KubeApiError;
+impl AgentManagment for OperatorApi {
```

Comment:
Do we have to implement AgentManagement for OperatorApi? It doesn't actually manage agents.